pkg/asset/store.go

```diff
@@ -14,47 +20,212 @@ type Store interface {
+		found, err = loadable.Load(LoadFileWith(s.onDiskFiles))
+		if err != nil {
+			return false, errors.Wrapf(err, "unexpected error when loading asset %s", asset.Name())
+		}
 	}
-	if s.assets == nil {
-		s.assets = make(map[reflect.Type]Asset)
+
+	// Try to load from on-disk files first.
+	switch {
+	case anyDirty && found:
+		return false, fmt.Errorf("cannot resolve, both parent assets and current asset %q are dirty", asset.Name())
+	case anyDirty:
+		dirty = true
+		logrus.Debugf("%generating %s...", indent, asset.Name())
+		if err := asset.Generate(parents); err != nil {
+			return false, errors.Wrapf(err, "failed to generate asset %s", asset.Name())
+		}
+	case found:
+		dirty = true
+		logrus.Debugf("%sFound %s on disk...", indent, asset.Name())
+	default:
+		// TODO(yifan): Try to load the asset from state file.
+		// If not exist, then generate using its parents.
+		logrus.Debugf("%sGenerating %s...", indent, asset.Name())
+		if err := asset.Generate(parents); err != nil {
+			return false, errors.Wrapf(err, "failed to generate asset %s", asset.Name())
+		}
 	}
+
 	s.assets[reflect.TypeOf(asset)] = asset
-	return nil
+	s.dirtyAssets[reflect.TypeOf(asset)] = dirty
+	return dirty, nil
+}
+
+// readAllFiles returns a map of filename path (relative to the clusterDir)
+// to the state
+func readAllFiles(clusterDir string) (map[string][]byte, error) {
+	fileMap := make(map[string][]byte)
+
+	// Don't bother if the clusterDir is not created yet because that
+	// means there's no assets generated yet.
+	_, err := os.Stat(clusterDir)
+	if err != nil && os.IsNotExist(err) {
+		return nil, nil
+	}
+
+	if err := filepath.Walk(clusterDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		filename, err := filepath.Rel(clusterDir, path)
+		if err != nil {
+			return err
+		}
+
+		data, err := ioutil.ReadFile(path)
+		if err != nil {
+			return err
+		}
+
+		fileMap[filename] = data
+
+		return nil
+
+	}); err != nil {
+		return nil, err
+	}
+	return fileMap, nil
+}
+
+// ByName will implement the Sort interface on the file list.
+// it will sort it by the filename in lexicographic order.
+type ByName []*File
+
+// Len returns the lens of the file list.
+func (b ByName) Len() int {
+	return len(b)
+}
+
+// Swap swaps the order of files[i] and files[j].
+func (b ByName) Swap(i, j int) {
+	b[i], b[j] = b[j], b[i]
+}
+
+// Less returns true if files[i]'s name is less than files[j]'s name lexicographically.
+func (b ByName) Less(i, j int) bool {
+	return b[i].Filename < b[j].Filename
+}
+
+// LoadFileWith returns a function which returns files
+// that matches the given input.
+func LoadFileWith(fileMap map[string][]byte) FileFetcher {
+	return func(input interface{}) ([]*File, error) {
+		var files []*File
+
+		switch input.(type) {
+		case string:
+			filename := input.(string)
+			data, ok := fileMap[input.(string)]
+			if ok {
+				files = append(files, &File{
```

Comment:
1) The flow follows a pattern where we return prematurely when a condition is encountered that prevents us from proceeding further. If the file does not exist, we return immediately. Otherwise, we are in the happy path where we create a `File`.
2) We can limit the scope of variables by not using the `files` variable. There is no need to store something in a local variable to then return that variable on the next line.

Reply:
To be clear, this is just a personal nit. Feel free to ignore it.